aws/rust-runtime/aws-runtime/src/auth/signing_application.rs

```diff
@@ -0,0 +1,277 @@
+use aws_smithy_runtime_api::box_error::BoxError;
+use aws_smithy_runtime_api::client::orchestrator::HttpRequest;
+
+/// A builder that configures fields for signing and, upon calling `.finalize()`, produces a
+/// signature and adds signing headers to the given request
+///
+/// # Examples
+/// ```rust,no_run
+/// use aws_runtime::auth::signing_application::SigningApplication;
+/// use aws_sigv4::http_request::SigningParams;
+/// # use aws_smithy_runtime_api::box_error::BoxError;
+/// use aws_smithy_runtime_api::client::orchestrator::HttpRequest;
+///
+/// fn signing_params<'a>() -> SigningParams<'a> {
+///     // --snip--
+/// #    todo!()
+/// }
+///
+/// fn request() -> HttpRequest {
+///     // --snip--
+/// #   HttpRequest::empty()
+/// }
+///
+/// # fn example() -> Result<(), BoxError> {
+/// let mut request = request();
+/// let signing_params = signing_params();
+/// SigningApplication::default()
+///     .request(&mut request)
+///     .signing_params(&signing_params)
+///     .finalize()
+/// # }
+/// ```
+#[derive(Debug, Default)]
+pub struct SigningApplication<'a> {
+    request: Option<&'a mut HttpRequest>,
+    signing_params: Option<&'a SigningParams<'a>>,
+    payload_override: Option<SignableBody<'a>>,
+    #[cfg(feature = "event-stream")]
+    deferred_signer_settings: Option<event_stream::DeferredSignerSettings<'a>>,
+}
+
+impl<'a> SigningApplication<'a> {
+    /// Sets the request (required).
+    pub fn request(mut self, request: &'a mut HttpRequest) -> Self {
+        self.set_request(Some(request));
+        self
+    }
+
+    /// Sets the request (required).
+    pub fn set_request(&mut self, request: Option<&'a mut HttpRequest>) -> &mut Self {
+        self.request = request;
+        self
+    }
+
+    /// Sets the [`SigningParams`] for the builder (required).
+    pub fn signing_params(mut self, signing_params: &'a SigningParams<'_>) -> Self {
+        self.set_signing_params(Some(signing_params));
+        self
+    }
+
+    /// Sets the [`SigningParams`] for the builder (required).
+    pub fn set_signing_params(
+        &mut self,
+        signing_params: Option<&'a SigningParams<'_>>,
+    ) -> &mut Self {
+        self.signing_params = signing_params;
+        self
+    }
+
+    /// Sets the [`SignableBody`] for the builder (optional).
+    pub fn payload_override(mut self, payload_override: SignableBody<'a>) -> Self {
+        self.set_payload_override(Some(payload_override));
+        self
+    }
+
+    /// Sets the [`SignableBody`] for the builder (optional).
+    pub fn set_payload_override(
+        &mut self,
+        payload_override: Option<SignableBody<'a>>,
+    ) -> &mut Self {
+        self.payload_override = payload_override;
+        self
+    }
+
+    #[cfg(feature = "event-stream")]
+    /// Sets the [`DeferredSignerSettings`](event_stream::DeferredSignerSettings) for the event stream signer (optional).
+    pub fn deferred_signer_settings(
```

Comment:
Can all this event stream deferred signer stuff be made `pub(crate)`? Customers shouldn't need any of this when signing directly.

Reply:
Makes sense. Addressed in a81ab9b.